Add BBLibc to the project

ronar opened the pull request.

ronar committed:
Add CMake project for BBLibc

ronar committed:
Refactor the project to use BBLibc library instead of external one

Remove lib/VC/BBLibc.def, lib/VC/BBLibc.lib

smartblade commented on src/BBLibc/BBLibc.cpp:
```diff
@@ -0,0 +1,88 @@
+#include <crtdbg.h>
+#include "direct.h"
```
Prefer `<...>` brackets here.

smartblade reviewed (changes requested):
You have done great work! I have added some notes.

ronar committed:
Fix B_Name::CompareNoCase()

ronar committed:
Fix includes order, remove unnecessary line in 'ObjExt.cpp'

ronar committed:
Fix variable name in B_IDataFile::Eof()

ronar commented on src/BBLibc/IDataFile.cpp:
```diff
@@ -0,0 +1,398 @@
+*/
+
+int B_IDataFile::Eof() {
+    long ecx = filePosition + cacheBlockFlag->filePosition;
```
Made `unsigned int`

ronar committed:
Remove BBLibc from `BOD_COMMON_INCLUDE_DIR` list

ronar committed:
Fix includes order of 'EntityFuncs.cpp'

ronar committed:
Remove BBLibc from 'BladeExt' dependencies

ronar committed:
Fix: use forward declaration instead of full header include of 'BLights.h' and 'RasterDevice.h'

ronar committed:
Fix: include standard library header <direct.h>

smartblade commented on src/Blade/ObjExt.cpp:
```diff
@@ -1615,14 +1623,14 @@ B_BitMap *RM_GetResource(const char *name)
     if (resource != NULL)
     {
-        B_BitMap *bitMap = new B_BitMap((B_BitMap &)resource->data);
+        B_BitMap *bitMap = new B_BitMap((B_BitMap &)resource->file_name);
```
What the reason of renaming `data` to `file_name`? It's not actually file name at this point.

ronar committed:
Remove unneeded dependency from CMakeLists.txt

ronar committed:
Fix includes of 'Control/anim.h'

ronar committed:
Remove Raster.dll dependency on BBlibc

ronar committed:
Return `#pragma once` to 'RasterDevice.h'

ronar committed:
Remove rOpenGL3.dll dependency on python15

ronar committed:
Return `#pragma once` to 'ROpenGLDeviceDLL.h'

ronar committed:
Fix: use forward declaration instead of full header include in 'ROpenGLDeviceDLL.h'

ronar committed:
Revert "Remove rOpenGL3.dll dependency on python15"

This reverts commit fa2940bbe7f98a9a2fe24c237ce577e79aee2e6f.

ronar commented on src/Raster/rOpenGL3/CMakeLists.txt:
```diff
@@ -39,6 +39,7 @@ target_link_libraries(
     AnalyticGeometry
     BBLibc
     Raster
+    python15
```
It's not compiling without it.

ronar committed:
Remove placeholder for B_Widget::SetSize(float w, float h)

ronar commented on src/BBLibc/BBLibc.cpp:
```diff
@@ -0,0 +1,88 @@
+#include "direct.h"
+
+#include "MessageManager.h"
+#define BBLIBC_LIB_EXPORT
```
If I put `#include "MessageManager.h"` below the `BBLIBC_LIB_EXPORT` it won't compile. Because `class B_MessageManager mout" (?mout@@3VB_MessageManager@@A)` already defined in the `"MessageManager.cpp"`

ronar replied:
And even `#include "direct.h"` should go below?

smartblade replied:
I see.

ronar committed:
Remove unneeded declaration

ronar committed:
Remove duplicated CallPythonObject declaration, move definition to 'BBLibc.cpp'

ronar commented on src/AnalyticGeometry/Vector.h:
```diff
@@ -2,8 +2,8 @@
-class B_IDataFile;
-class B_ODataFile;
+#include "IDataFile.h"
+#include "ODataFile.h"
```
Otherwise it's not working. I presume because these are defined in `Vector.cpp`:
```
/*
* Module:                 Blade.exe
* Entry point:            0x0040ACDD
* VC++ mangling:          ??5@YAAAVB_IDataFile@@AAV0@AAVB_Vector@@@Z
*/

B_IDataFile& operator >>(B_IDataFile& file, B_Vector &vector)
{
    file.Read(&vector, sizeof(B_Vector));
    return file;
}


/*
* Module:                 Blade.exe
* Entry point:            0x0040ACF4
* VC++ mangling:          ??6@YAAAVB_ODataFile@@AAV0@AAVB_Vector@@@Z
*/

B_ODataFile& operator <<(B_ODataFile& file, B_Vector &vector)
{
    file.Write(&vector, sizeof(B_Vector));
    return file;
}
```
They're being imported from the `BBLibc` instead

smartblade replied:
OK

ronar replied:
I think we might move it to `IDataFile.cpp` file

smartblade replied:
No, they originally belongs to `Vector.cpp`(`Blade.exe`). `IDataFile.cpp` is at `BBLibc.dll`

ronar committed:
Update OutputWin32Error function: FormatMessageA -> FormatMessage

ronar committed:
Move B_Pal class to separate file

ronar commented on src/Python-1.5.2/CMakeLists.txt:
```diff
@@ -30,7 +30,7 @@ if (MSVC)
     set_target_properties(
         ${TARGET} PROPERTIES
-        LINK_FLAGS "/NODEFAULTLIB:libc.lib"
+        LINK_FLAGS "/NODEFAULTLIB:libc.lib /SAFESEH:NO"
```
If open a project as Visual Studio project, it's not compiling without it.

ronar commented on src/BBLibc/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+target_include_directories(${TARGET} PRIVATE ${BOD_COMMON_INCLUDE_DIR})
+target_link_directories(${TARGET} PRIVATE ${BOD_EXTERNAL_LIB_DIR})
+# Disable SAFESEH
+target_link_options(${TARGET} PRIVATE "/SAFESEH:NO")
```
If open a project as Visual Studio project, it's not compiling without it.

ronar commented on src/BBLibc/BBLibc.cpp:
```diff
@@ -0,0 +1,88 @@
+*/
+
+const char* GetConfigDirectory() {
+    // TODO implement
```
Will do this last. Let's first resolve other issues

smartblade commented on src/Raster/rOpenGL3/RTrisDeviceDLL.h:
```diff
@@ -1,4 +1,7 @@
+#ifndef R_TRIS_DEVICE_DLL_H
+
+#define R_TRIS_DEVICE_DLL_H
+
```
Still unresolved

smartblade commented on src/Raster/RasterDevice.h:
```diff
@@ -1,21 +1,25 @@
-class B_Color;
 class B_Font;
-struct B_Particle;
-class B_ParticleGType;
```
Still unresolved.

ronar replied:
That's strange. Because it has it this commit: #[ed41cfa](https://github.com/smartblade/BldMystery/pull/5/commits/ed41cfa8d18e357a02748a2bcdb0c00ebf9277be)

smartblade replied:
Also `Name`, `Color` and `Bitmap24`.

ronar replied:
Understood.

ronar replied:
`B_Color` was needed, without it, it does not compile. So I left it.

smartblade commented on src/BBLibc/BBLibc.h:
```diff
@@ -0,0 +1,39 @@
+#define BBLIBC_H
+
+#include <bld_system.h>
+#include <bld_python.h>
```
I suggest to remove Python include and use forward declaration `struct PyObject`. It can resolve https://github.com/smartblade/BldMystery/pull/5#discussion_r1755127785.

ronar committed:
Return `#pragma once` to 'RTrisDeviceDLL.h'

ronar committed:
Fix entry point for B_Pal::GetBlue

ronar committed:
Remove placeholder for B_Widget::Draw

ronar committed:
Remove rOpenGL3.dll dependency on python15 (2)

Change headers order of 'BBLibc.cpp' and 'ROpenGLDeviceDLL.cpp'

ronar committed:
Remove `/SAFESEH:NO` flag

ronar committed:
Fix: use forward declarations instead of full header include (B_BitMap24, B_Name) in 'RasterDevice.h'

ronar committed:
Correct B_Resource field (file_name -> data), revert changes of 'ObjExt.cpp'

ronar committed:
Mark unimplemented functions for 'process_asm.py'

Change order of some functions to match the original offsets
Refactor field names to include offset
Add new functions:
B_List::B_List(B_List const &)
B_ResourceManager::B_ResourceManager()
B_ResourceManager::~B_ResourceManager()
B_ResourceManager* B_ResourceManager::operator =(B_ResourceManager*)
B_IDataFile& operator >>(B_IDataFile&, char&)
B_IDataFile& operator >>(B_IDataFile&, unsigned short&)
B_IDataFile& operator >>(B_IDataFile&, unsigned long&)
B_IDataFile& operator >>(B_IDataFile&, B_BitMap&)
B_ODataFile& operator <<(B_ODataFile&, char const&)
B_ODataFile& operator <<(B_ODataFile&, unsigned short const&)
B_ODataFile& operator <<(B_ODataFile&, long const&)
B_ODataFile& operator <<(B_ODataFile&, unsigned intconst&)
B_ODataFile& operator <<(B_ODataFile&, B_BitMap const&)
B_ODataFile& operator <<(B_ODataFile&, B_BitMap24 const&)
B_ObjDscr::B_ObjDscr()
B_ObjDscr::~B_ObjDscr()
Remove:
int B_ParticleGType::operator >>(B_IDataFile&)
int B_ParticleGType::operator <<(B_IDataFile&)
Code style fixes

smartblade commented on src/BBLibc/MessageChannel.h:
```diff
@@ -14,10 +14,11 @@
 
 
-class LIB_EXP B_MessageChannel : public B_NamedObj {
+class LIB_EXP B_MessageChannel : public B_Name
```
No, it was right. B_MessageChannel is the child of `B_NamedObj`, not `B_Name`.

ronar committed:
Revert B_MessageChannel to inherit from B_NamedObj, code style fix

ronar committed:
Fix B_Name::B_Name(const char*)

smartblade reviewed (approved):
Looks good now. Thanks!
